Add <input type=checkbox switch> infrastructure

annevk opened the pull request:
#### 9ea9298663a4c17a15be6c8eae83a67f8d345f9f
<pre>
Add &lt;input type=checkbox switch&gt; infrastructure
<a href="https://bugs.webkit.org/show_bug.cgi?id=262966">https://bugs.webkit.org/show_bug.cgi?id=262966</a>
rdar://116746765

Reviewed by Aditya Keerthi.

Adds the infrastructure needed to eventually be able to support an HTML
switch control. It notably does not cover any rendering aspects (except
for not rendering a checkbox).

This feature is documented in <a href="https://github.com/whatwg/html/pull/9546.">https://github.com/whatwg/html/pull/9546.</a>

This builds on excellent prototyping work done by Lily Spiniolas.

This commit includes some fixes resulting from running
update-webkit-localizable-strings.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::checkboxOrRadioValue const):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::valueMissingText const):
(WebCore::CheckboxInputType::attributeChanged):
(WebCore::CheckboxInputType::shouldAppearIndeterminate const):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isSwitch const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::validationMessageValueMissingForCheckboxText):
(WebCore::validationMessageValueMissingForSwitchText):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/style/StyleAppearance.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/style/StyleAppearance.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::encode):
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView fullScreenWindowSceneDimmingAction]):

Canonical link: <a href="https://commits.webkit.org/269247@main">https://commits.webkit.org/269247@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1e2661c91a80393c44afa81459fd3525bccbd2ba

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22024 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23093 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23907 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20388 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22270 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26484 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22553 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21452 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22255 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21909 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24760 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19018 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19961 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26218 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19144 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20039 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20186 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24085 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21397 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20615 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/17551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25440 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19966 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6000 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5248 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24174 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/26715 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20565 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5832 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/CheckboxInputType.cpp:
```diff
@@ -100,7 +110,7 @@ bool CheckboxInputType::matchesIndeterminatePseudoClass() const
 {
     ASSERT(element());
-    return element()->indeterminate();
+    return element()->indeterminate() && !element()->isSwitch();
```
Just to confirm my understanding here for why `indeterminate()` doesn't encapsulate `isSwitch()`: Do we still want to reflect `indeterminate` true/false through JS, but not for rendering?

annevk replied:
Yeah, `indeterminate` state can always be changed regardless of the control. It's a bit weird, but it seems better to be consistent with the weirdness.

annevk committed:
Add <input type=checkbox switch> infrastructure
https://bugs.webkit.org/show_bug.cgi?id=262966
rdar://116746765

Reviewed by Aditya Keerthi.

Adds the infrastructure needed to eventually be able to support an HTML
switch control. It notably does not cover any rendering aspects (except
for not rendering a checkbox).

This feature is documented in https://github.com/whatwg/html/pull/9546.

This builds on excellent prototyping work done by Lily Spiniolas.

This commit includes some fixes resulting from running
update-webkit-localizable-strings.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineAccessibilityRoleFromNode const):
(WebCore::AccessibilityNodeObject::checkboxOrRadioValue const):
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::toCSSValueID):
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/CheckboxInputType.cpp:
(WebCore::CheckboxInputType::valueMissingText const):
(WebCore::CheckboxInputType::attributeChanged):
(WebCore::CheckboxInputType::shouldAppearIndeterminate const):
* Source/WebCore/html/CheckboxInputType.h:
* Source/WebCore/html/HTMLAttributeNames.in:
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::isSwitch const):
* Source/WebCore/html/HTMLInputElement.h:
* Source/WebCore/html/HTMLInputElement.idl:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::validationMessageValueMissingForCheckboxText):
(WebCore::validationMessageValueMissingForSwitchText):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/style/StyleAppearance.cpp:
(WebCore::operator<<):
* Source/WebCore/style/StyleAppearance.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::encode):
(IPC::ArgumentCoder<ControlPart>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView fullScreenWindowSceneDimmingAction]):

Canonical link: https://commits.webkit.org/269247@main
